Transient stats for SQL statements.

cuongdo opened the pull request:
Added support for counts of various SQL statements and bytes in/out for the pgwire protocol.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4031)

<!-- Reviewable:end -->

Commit:
Transient stats for SQL statements. Also restructured the way the Registry instances interact
with one another.

cuongdo commented:
CC @maxlang for the front-end code, since I copied and pasted

petermattis commented:
LGTM, though I didn't look at the javascript.

---

Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CuQBP1YETazKiOYMz:-K9CuQBP1YETazKiOYN-:-2045797847) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/testserver.go#L334)):*
Ah, I was wondering why the returned map had floats for the counts. I wonder if it would be better to provide a method to get a particular stat:

```
func (ts *TestServer) GetCounter(name string) (*metric.Counter, error) { 
  var c *metric.Counter
  var err error
  ts.Server.metaRegistry.Each(func(n string, v interface{}) {
    if c == nil && err == nil name == n {
      var ok bool
      if c, ok = v.(*metric.Counter); !ok {
        err = util.Errorf("%s is not a counter: %T", n, v)
      }
    }
  })
  return c, err
}
```

---

*[sql/executor.go, line 520 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CrmECO6SYD5UjviVw:-K9CrmECO6SYD5UjviVx:334390269) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/executor.go#L520)):*
What about non-DDL, non-DML statements (e.g. `SET`, `SHOW`, etc)? Even if we don't display the counts, perhaps capture these operations in `miscCount` stat.

@mtracy: Should we keep separate stats for every operation and aggregate the DDL stats for display in the UI?

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9Csat0HwItgmxeyuNZ:90063840) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
We generally do not capitalize error messages.

---

*[sql/pgwire/encoding.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Ctusblb0iC-f29Qo3:-K9Ctusc-jR816jMHHhq:989856114) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/encoding.go#L145)):*
I'm finding this field a bit confusing. Can't we increment the `bytesOutCount` stat whenever we write data to the connection, regardless of whether it has been flushed? That is, I'm thinking you should either pass a `*metric.Counter` to `writeBuffer.finishMsg` or have `finishMsg` return a count of the number of bytes written.

---

*[util/metric/README.md, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CqXRlKev09L2JHfAr:-K9CqXRlKev09L2JHfAs:-2012143220) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L1)):*
Nice docs. I wonder if this should be moved to `doc.go` (i.e. a package document comment). See `sql/doc.go` for an example. 

---

*[util/metric/README.md, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Cqm-XVtpmy5jiT2VE:-K9Cqm-YGT385BH63AhB:-1436756554) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L38)):*
This only works in insecure mode, but that's ok since the developer is probably running `--dev`. Might want to note that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-:-K9Cv6xRkspyvG7rxuZE:-466286991)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 12 of 16 files at r1.
Review status: 12 of 16 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CuQBP1YETazKiOYMz:-K9CzmEjBCDht_ELGkof:446987583) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/testserver.go#L334)):*
Even if you don't go this route, marshalling to JSON and back is really a big hammer. Can you perhaps construct the map yourself using `Each()`?

---

*[sql/metric_test.go, line 75 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-EV6Zebt4RL7zUaC:-K9D-EV7lc9QuoeznVCZ:-1050375080) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L75)):*
this is nice.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9D-it-H7_XfVcykfi_:180058276) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
should this come after the error check?

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9D-n6SJ7KP0q35ZS3y:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
ditto

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 16 files at r1.
Review status: 15 of 17 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9PlqRvUaywohbYoKXV:-K9PlqRwk0PGdw1CQ9pX:73662936) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/server.go#L84)):*
can you add a comment explaining why two registries are passed in? It seems to me that this is only used to add the `sql` family of metrics in `Executor`, which might as well receive the `metaRegistry` and register a subregistry with the appropriate metrics (and tests would only need the `metaRegistry` to extract the relevant submetrics using `.Each`). Anticipating the addition of more metrics families, that seems to be the more elegant pattern.

---

*[server/status.go, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9PmbmD5yf7CinIh66F:-K9PmbmEi4rMeYMSFUy7:-1440377564) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/status.go#L95)):*
Are we planning for per-store metrics?

---

*[util/metric/README.md, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CqXRlKev09L2JHfAr:-K9PotSDc0r9t8j7f_EL:-388288417) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L1)):*
There's also the question of whether this is the right place for this information (and whether it should show up in godoc). As it is, nothing in the `util/metric` package is specific to how we use it within cockroachdb. If this is going to be the place where contributors look for this information, I'm in favor of keeping it here. Naively, the "natural" point where I would have looked for it is the `metaRegistry` field on Server (or perhaps in the `server` docs), but obviously I'm biased by prior knowledge.

---

*[util/metric/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Po64F3xvZc8Bqbsn6:-K9Po64F3xvZc8Bqbsn7:-1061231454) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L17)):*
referring to my other comment, the code here would be something like

```go
registry := ...
exec := &Executor{
  selectCount: registry.Counter("select.count"),
  ...
}
metaRegistry.Add("sql.%s", registry)
```

which avoids passing this extra registry around and also avoids the repetitive mention of the `sql` prefix.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 15 of 17 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CuQBP1YETazKiOYMz:-K9QKawKCDzJk2d1ucre:-100192730) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/testserver.go#L334)):*
+1 on using `Each()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 15 of 17 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ui/ts/pages/nodes.ts, line 415 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9TbKunsMrJ5NfNLSVy:-K9TbKuoaegX_YPbOHJX:81899413) ([raw file](https://github.com/cockroachdb/cockroach/blob/c199393c0471677622e4060b7c5de2a694278e19/ui/ts/pages/nodes.ts#L415)):*
LGTM for now, though these graphs will probably need to be moved to a different tab/page.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 1 of 19 files reviewed at latest revision, 13 unresolved discussions.

---

*[server/server.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9PlqRvUaywohbYoKXV:-K9XTJziFqjgIacOKUGg:-1215643127) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/server.go#L84)):*
I changed a lot of things about how the various `Registry` instances work. The hierarchy should hopefully be a lot clearer now. 

---

*[server/status.go, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9PmbmD5yf7CinIh66F:-K9SN01dtMHzikG54WgP:1272160628) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/status.go#L95)):*
Not that I know of, though that can certainly be added.

---

*[server/testserver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CuQBP1YETazKiOYMz:-K9XTbWMPk-LDglitQQh:-69480795) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/server/testserver.go#L334)):*
OK, I made it a lot cleaner. I chose to implement `MustGetCounter()` instead of cloning the whole map, because some of the metric types (e.g. the histogram) do not yet have a method to clone them.

---

*[sql/executor.go, line 520 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CrmECO6SYD5UjviVw:-K9XUFdsW2F9UNf_N89f:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/executor.go#L520)):*
Done.

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9XWnIyb_4qmaUX5SYa:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
Done.

---

*[sql/pgwire/encoding.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Ctusblb0iC-f29Qo3:-K9XbOPhQrQpKJRHT1qu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/encoding.go#L145)):*
Done.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9XbQCtCVAUa2_ajtiY:1395323853) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
I've removed this code.

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9XbSraPmZJ9dQugGQ0:1395323853) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
I've removed this code.

---

*[util/metric/README.md, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9CqXRlKev09L2JHfAr:-K9YI7qLdV5Qp6lHg57l:-1110747165) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L1)):*
I see both of your points, so I moved the docs to `util/metric/doc.go` and added a pointer to that in the comment for the `registry` field of `Server`.

---

*[util/metric/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Po64F3xvZc8Bqbsn6:-K9XmrR3WR0mjO8hvFVg:-800241810) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L17)):*
This is fixed with my Registry cleanup.

---

*[util/metric/README.md, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Cqm-XVtpmy5jiT2VE:-K9Xn8vwXc4GOqElupSy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/util/metric/README.md#L38)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 19 files at r4.
Review status: 3 of 19 files reviewed at latest revision, 14 unresolved discussions.

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9YUp6WTj40ca8AabPZ:-1105124553) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
is there a guarantee on name uniqueness? doesn't seem like you need `found`.

---

*[server/testserver.go, line 339 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YV-gqkPp1Kr8TKGUw:-K9YV-gqkPp1Kr8TKGUx:44287828) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L339)):*
`err` appears to be unused.

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9YV4wkISUP0eUe0lYt:-1057464555) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
Doesn't appear to have changed.

---

*[sql/pgwire/encoding.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Ctusblb0iC-f29Qo3:-K9YVBFwpB9y1jWbRAtp:-585609667) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/encoding.go#L145)):*
Hm, not seeing it.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9YUT0j21zcH6kbPIPm:1744993208) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
Are you sure?

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9YUdSthNKkDRAiA1yf:1744993208) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
Are you sure?

---

*[util/metric/README.md, line 0 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUJsKrWpREYDPCHqK:-K9YUJsLyWHQwZ5sPfvy:-1530693521) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/README.md#L0)):*
Is there still a file here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 3 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9YXJ2sH6P_XcKSPCRK:-466296891) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
Yes, there is. `found` is mainly used to short-circuit the rest of the code in the anonymous function passed to `Each()`after we've found the metric we're looking for.

---

*[server/testserver.go, line 339 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YV-gqkPp1Kr8TKGUw:-K9YXZwJcgP98XbJM3VZ:-2146338981) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L339)):*
It's used like `found` is above.

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9YXjngGphamE0_MLfr:2138239997) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
Ah right, fixed the other messages but not this one. Done.

---

*[sql/pgwire/encoding.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Ctusblb0iC-f29Qo3:-K9YXpZLca_Axhxok6l_:47894784) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/encoding.go#L145)):*
I added `bytecount` to `writeBuffer`, replacing a different field (`nwritten`). I wonder if my method of squashing commits overwrote some state here, because the previous revision did look different.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9YYDZj3ZcOUHkXCD9V:-1521591661) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
Ah you're right. Moved the code after the error.

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9YYK5-rmKYfjW00PN4:2089799337) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
Ah, I misread your comment earlier. Moved the code. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 3 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9YajL85_ysSpH-beUP:1048769749) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
Is that worth it? This is short-circuiting a single type assertion.

---

*[server/testserver.go, line 339 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YV-gqkPp1Kr8TKGUw:-K9YafFMuIEMTCC3D0PV:-226649111) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L339)):*
What's the point?

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9YatFA6vZsBbw8FsLm:984154116) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
Still not changed.

---

*[sql/pgwire/encoding.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Ctusblb0iC-f29Qo3:-K9YaxfMt3tucbyYUyDh:2022453011) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/encoding.go#L145)):*
I was always looking at the total diff, so I doubt there's an issue with history.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9Yb4ga36MuymMxMbiM:892327448) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
Doesn't look like it.

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9Yb6P80qjVvWGlpsva:892327448) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
Doesn't look like it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 3 of 19 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-:-K9Ybs7PQBVNX-OzRV8J:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 16 files at r1, 19 of 19 files at r4.
Review status: 17 of 19 files reviewed at latest revision, 27 unresolved discussions.

---

*[server/server.go, line 85 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YTInSinSqzRpfjTa6:-K9YTInTL83_1DZ-Uj1p:2147003519) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/server.go#L85)):*
s/Add/Adding/

---

*[server/status/monitor.go, line 50 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YY4m-D0uWxh5CKbEm:-K9YY4m-D0uWxh5CKbEn:-898363780) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L50)):*
add something to the effect of:

> Upon receiving notice of a NodeID having been set (by means of a OnStartNode event) , the given subregistries are added to serverRegistry in timeseries-compatible format, along with metrics tracked within the created NodeStatusMonitor.

---

*[server/status/monitor.go, line 53 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUvJWsWu8iK2LgXDP:-K9YUvJWsWu8iK2LgXDQ:1101058475) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L53)):*
The amount of registries flying around here is pretty high (thanks to the late birth of `NodeID`).
Since this creates a new registry only for the prefix, I think some confusion can be eliminated by adding straight to `nodeRegistry` and spelling out the `exec` prefix. That removes at least `execRegistry`. Your call though.

---

*[server/status/monitor.go, line 212 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUeU6Vp2rbphbONjp:-K9YUeU6Vp2rbphbONjq:1580032170) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L212)):*
s/all sub-registries/the node registry/

---

*[server/testserver.go, line 328 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YVwiEimfuEMhxEX0J:-K9YVwiEimfuEMhxEX0K:130787888) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L328)):*
if this is called `MustGetCounter` I would expect a panic if there's no such counter.

---

*[server/testserver.go, line 331 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YVWnWs7SW6VaqAVE0:-K9YVWnXSMmyEDUNGzWN:571681241) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L331)):*
nit: `var found bool`

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9YVn_CsSr9paWIjLaw:-1194123660) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
it's internally a map, so names are unique. I think since this is for testing and is allowed to panic, something like

```go
ts.registry.Each(func(n string, v interface{}) {
  if name == n {
    return v.(*metric.Counter).Count()
  }
}
panic(fmt.Sprintf("metric %s not found", name))
```

is good enough.

---

*[sql/executor.go, line 493 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YZESopfL-qgCx3UGp:-K9YZESopfL-qgCx3UGq:952097859) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L493)):*
should the `txnCount` metric be something like `beginTxnCount`? That makes it more self-evident what it's counting (otherwise you might worry about errors, in particular transaction retries, which btw I would also love to see counts for; doesn't have to be now though).

---

*[sql/executor.go, line 532 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YZieGjxkdAq6DBK1Z:-K9YZieHjhIy0jsn1nI2:129708325) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L532)):*
those newlines look spurious to me, but that's just my OCD.

---

*[sql/executor.go, line 552 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9YWEPeKz_Sngq8NKGy:-1821562944) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552)):*
the comment is overly specific (it refers other packages in the way that we use it, but not necessarily have to or will forever). It's helpful, but consider something like

> Registry returns a registry with all metrics tracked by this executor registered to it, which can be used to access its stats (or can be added to a higher-level registry).

---

*[sql/pgwire/encoding.go, line 68 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Y_3DphJ4jkKbweAxM:-K9Y_3Dq1jmchGPm3ji9:1517870974) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/encoding.go#L68)):*
explain the returned `int`. In particular, as it looks like below, it can be intentionally nonzero when an error is returned. That's somewhat unusual in Go, so the comment should mention it.

---

*[sql/pgwire/encoding.go, line 87 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Y_FeOBcLgMeMvlsL4:-K9Y_FeOBcLgMeMvlsL5:1660101171) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/encoding.go#L87)):*
explain parameter.

---

*[sql/pgwire/server.go, line 56 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YajdtyPM2EacJnf8N:-K9YajdvM6YDt6IeAg0m:1930751168) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L56)):*
do you need this type anywhere? If you just want the grouping,

```go
type Server struct {
  // ...
  metrics struct {
    bytesInCount  *metric.Counter
    bytesOutCount *metric.Counter
  }
}
```

---

*[sql/pgwire/server.go, line 63 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbG20VEWcU1iOAdq2:-K9YbG20VEWcU1iOAdq3:-892659931) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L63)):*
no main registry here as far as I can tell.

---

*[sql/pgwire/server.go, line 228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9YX-QL3bBmeRS3FWB7:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228)):*
ditto.

---

*[sql/pgwire/v3.go, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbchYlmhUS1ZoH-6e:-K9YbchYlmhUS1ZoH-6f:624515779) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/v3.go#L126)):*
it's weird that a new connection gets a writeBuffer whose byteCount doesn't start at zero. Maybe that's the most ergonomic way, but add a comment somewhere to justify that choice.

---

*[util/metric/doc.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXHqISWlzgNgWCmdb:-K9YXHqISWlzgNgWCmdc:-105762904) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L34)):*
`sql.select.count`

---

*[util/metric/doc.go, line 45 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXO_0rwZbAdWMnyXr:-K9YXO_0rwZbAdWMnyXs:332526874) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L45)):*
Take a look at the output of this in `godoc`. I think you may want `## Sub-registries` or something like that, but haven't checked.

---

*[util/metric/doc.go, line 81 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXdkAfpFMB-SiiOsO:-K9YXdkAfpFMB-SiiOsP:-1224788526) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L81)):*
spurious newlines.

---

*[util/metric/registry.go, line 160 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbvOWKaZ01KtFbq5A:-K9YbvOWKaZ01KtFbq5B:2027692775) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/registry.go#L160)):*
As in other places, I prefer comments that stand for themselves, without relating to other parts of our use case which can be confusing for others.

maybe `SubRegistry` should move closer to `Node`. This `NodeID`-thing is hopefully the only time we run into this issue (ok, maybe again with `Store` at some point). Fine to leave it here for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM on the basic registry handling (the NodeID stuff makes it hard to be elegant). Someone else should green-light the pgwire byte counting logic, but other than that this looks like a good end of this iteration.

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

*[sql/metric_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Csat0HwItgmxeyuNY:-K9PqQaCZSDT0eKoHJXU:-874532316) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/metric_test.go#L42)):*
fixed in the latest revision. Probably comment-before-push.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-:-K9YcshTlkslLSiyVCBj:-1196910396)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 7 of 19 files reviewed at latest revision, 27 unresolved discussions.

---

*[server/server.go, line 85 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YTInSinSqzRpfjTa6:-K9Yd5xSKK1qoH4FJ7br:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/server.go#L85)):*
Done.

---

*[server/status/monitor.go, line 50 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YY4m-D0uWxh5CKbEm:-K9Ydfggd09wDfbwqc0X:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L50)):*
Done.

---

*[server/status/monitor.go, line 53 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUvJWsWu8iK2LgXDP:-K9Ye3Y_AZ3T_kYPNq-J:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L53)):*
Done.

---

*[server/status/monitor.go, line 212 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUeU6Vp2rbphbONjp:-K9YeFdC3S0F_KsXTZ2c:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/status/monitor.go#L212)):*
Done.

---

*[server/testserver.go, line 328 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YVwiEimfuEMhxEX0J:-K9YeVtsNuo-0MRBPO8j:1979007644) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L328)):*
Panic added.

---

*[server/testserver.go, line 331 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YVWnWs7SW6VaqAVE0:-K9YeXqqWCfwNX0fFR4Q:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L331)):*
Done.

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9Yfi1EuRnTvLBKfGjQ:550128754) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
OK, I simplified the loop. `found` is still needed, because we need to panic if we haven't found the metric. The return can't go in the anonymous function passed into `Each`, because that function has no return value and can't return from `MustGetCounter` anyway.

---

*[server/testserver.go, line 339 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YV-gqkPp1Kr8TKGUw:-K9Yg0a-g_WNL_sZgNZ-:2106422686) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L339)):*
removed, and replaced with panics

---

*[sql/executor.go, line 493 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YZESopfL-qgCx3UGp:-K9Ynb_EN3-vdCZDFbgt:1540106479) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L493)):*
changed it to `txnBeginCount`

---

*[sql/executor.go, line 532 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YZieGjxkdAq6DBK1Z:-K9YgVcvBwC9-Aws_aSt:1921196252) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L532)):*
removed for the sake of consistency

---

*[sql/executor.go, line 552 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9YgtINEBnKK1nPaSoV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552)):*
Done.

---

*[sql/pgwire/encoding.go, line 68 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Y_3DphJ4jkKbweAxM:-K9YhopuDhVfJbYQWi1X:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/encoding.go#L68)):*
Done.

---

*[sql/pgwire/encoding.go, line 87 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9Y_FeOBcLgMeMvlsL4:-K9YiBsFY25fmlRy7ucq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/encoding.go#L87)):*
Done.

---

*[sql/pgwire/server.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-iszDH3nSAcStW8P:-K9YioG-SiPos3pyaJxO:-690605587) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L152)):*
Push after publish. Sorry.

---

*[sql/pgwire/server.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9D-n6SJ7KP0q35ZS3x:-K9Yj9cJBY-CDjq3X0e8:-611441595) ([raw file](https://github.com/cockroachdb/cockroach/blob/e522bdc1d997b91ea5cdddaa41ee5273a6b5b860/sql/pgwire/server.go#L182)):*
Publish before push

---

*[sql/pgwire/server.go, line 56 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YajdtyPM2EacJnf8N:-K9YibFzoEbRTJYs1nXW:1645189152) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L56)):*
It's passed into `makeV3Conn` below.

---

*[sql/pgwire/server.go, line 63 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbG20VEWcU1iOAdq2:-K9YimuZ1cne74K06FoI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L63)):*
Done.

---

*[sql/pgwire/server.go, line 228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9YjNRHIGZv3ZBHxOfL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228)):*
Done.

---

*[sql/pgwire/v3.go, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbchYlmhUS1ZoH-6e:-K9Yk3QwUvvjvlC1vkG7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/v3.go#L126)):*
Done.

---

*[util/metric/doc.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXHqISWlzgNgWCmdb:-K9YkABV5FolD8GeS556:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L34)):*
Done.

---

*[util/metric/doc.go, line 45 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXO_0rwZbAdWMnyXr:-K9YkKfiblOQZaIoaEzM:-57770247) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L45)):*
I did review the `godoc` output. Sub-registries is actually meant to be at the same level as "Adding a new metric" and "Testing." In any case, I can't find any subheading or Markdown-style functionality.

---

*[util/metric/doc.go, line 81 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YXdkAfpFMB-SiiOsO:-K9YkoNU8-MrgNO9sx_h:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/doc.go#L81)):*
Done.

---

*[util/metric/registry.go, line 160 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YbvOWKaZ01KtFbq5A:-K9YlsjrRkiMXf55lbiz:665214484) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/registry.go#L160)):*
moved to `status/monitor.go`

---

*[util/metric/README.md, line  \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUJsKrWpREYDPCHqK:-K9Ycp57u9agsisSx-dL:435979010) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/util/metric/README.md#L)):*
No, it was moved to util/metric/doc.go and reformatted appropriately.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 12 of 12 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[server/testserver.go, line 334 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YUp6VdbjqcHA6Q-1Q:-K9Ytpy4kz2KyEUCitLB:1247155772) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/server/testserver.go#L334)):*
ah, yeah. Duh.

---

*[sql/executor.go, line 552 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9YuUe0Sjw3Zp4AvIfL:-476929764) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552)):*
s/or be added/or add it to/
s/returns registry/returns a registry/

---

*[sql/pgwire/server.go, line 228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9Yuqu-YhCO-Vhz_eOm:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228)):*
ditto.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/pgwire/server.go, line 227 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YyvPSjPwdfWkM18JS:-K9YyvPSjPwdfWkM18JT:1618789527) ([raw file](https://github.com/cockroachdb/cockroach/blob/156d5d4c1a673e989e9827e2ab33d812645ad044/sql/pgwire/server.go#L227)):*
missing article

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/pgwire/server.go, line 227 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YyvPSjPwdfWkM18JS:-K9YzGJzLdA5LgaX1t9k:-1484259237) ([raw file](https://github.com/cockroachdb/cockroach/blob/156d5d4c1a673e989e9827e2ab33d812645ad044/sql/pgwire/server.go#L227)):*
Oops. Disregard.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 17 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 552 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9bXaEf8T2EpwwOtYU_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552)):*
Done.

---

*[sql/pgwire/server.go, line 228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9bbObJffst_ouMHI50:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031)*
<!-- Sent from Reviewable.io -->

mtracy commented:
Hi all,

I'm not quite sure how I got on this mailing list; I think that there must
be someone on your team with a similar name/email address. That being said,
I am in fact a computer engineer by education (currently in graduate school
for my masters), and have quite an interest in database design, so after
looking into cockroachdb, I'm pretty interested. If possible, I'd like to
help you guys out. Any pointers on where I can look to get started?

-Mitchell Tracy

On Wed, Feb 3, 2016 at 9:51 AM, Cuong Do <notifications@github.com> wrote:

> Review status: 17 of 19 files reviewed at latest revision, 4 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *sql/executor.go, line 552 [r4]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9bXaEf8T2EpwwOtYU_:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552>):*
> Done.
> ------------------------------
>
> *sql/pgwire/server.go, line 228 [r4]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9bbObJffst_ouMHI50:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228>):*
> Done.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4031>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179274342>
> .
>

cuongdo commented:
@tamird: any additional comments on the pgwire byte counting code?

---

Review status: 17 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-:-K9byRoD2l4t77LJrcuq:-1009366746)*
<!-- Sent from Reviewable.io -->

JackKrupansky commented:
Looks like your user name of mtracy is easily confused with mrtracy.
Someone tried to mention mrtracy and mentioned you (mtracy) by accident
earlier in this thread.

Start with the low-level readme in the repo and then read the design doc:
https://github.com/cockroachdb/cockroach
https://github.com/cockroachdb/cockroach/blob/master/docs/design.md

And study everything on the web site as well
http://www.cockroachlabs.com/

Including the blog.


-- Jack Krupansky

On Wed, Feb 3, 2016 at 11:03 AM, mtracy <notifications@github.com> wrote:

> Hi all,
>
> I'm not quite sure how I got on this mailing list; I think that there must
> be someone on your team with a similar name/email address. That being said,
> I am in fact a computer engineer by education (currently in graduate school
> for my masters), and have quite an interest in database design, so after
> looking into cockroachdb, I'm pretty interested. If possible, I'd like to
> help you guys out. Any pointers on where I can look to get started?
>
> -Mitchell Tracy
>
> On Wed, Feb 3, 2016 at 9:51 AM, Cuong Do <notifications@github.com> wrote:
>
> > Review status: 17 of 19 files reviewed at latest revision, 4 unresolved
> > discussions, some commit checks failed.
> > ------------------------------
> >
> > *sql/executor.go, line 552 [r4]
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9bXaEf8T2EpwwOtYU_:-499146009
> >
> > (raw file
> > <
> https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552
> >):*
> > Done.
> > ------------------------------
> >
> > *sql/pgwire/server.go, line 228 [r4]
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9bbObJffst_ouMHI50:-499146009
> >
> > (raw file
> > <
> https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228
> >):*
> > Done.
> > ------------------------------
> >
> > *Comments from the review on Reviewable.io
> > <https://reviewable.io:443/reviews/cockroachdb/cockroach/4031>*
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179274342>
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179313567>
> .
>

tamird commented:
@cuongodo go ahead; I haven't kept up.

---

Review status: 17 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-:-K9cSyZCXNyGWLq3AqK9:998230116)*
<!-- Sent from Reviewable.io -->

mtracy commented:
Awesome thanks a lot!

-Mitchell Tracy

On Wed, Feb 3, 2016 at 11:41 AM, Jack Krupansky <notifications@github.com>
wrote:

> Looks like your user name of mtracy is easily confused with mrtracy.
> Someone tried to mention mrtracy and mentioned you (mtracy) by accident
> earlier in this thread.
>
> Start with the low-level readme in the repo and then read the design doc:
> https://github.com/cockroachdb/cockroach
> https://github.com/cockroachdb/cockroach/blob/master/docs/design.md
>
> And study everything on the web site as well
> http://www.cockroachlabs.com/
>
> Including the blog.
>
>
> -- Jack Krupansky
>
> On Wed, Feb 3, 2016 at 11:03 AM, mtracy <notifications@github.com> wrote:
>
> > Hi all,
> >
> > I'm not quite sure how I got on this mailing list; I think that there
> must
> > be someone on your team with a similar name/email address. That being
> said,
> > I am in fact a computer engineer by education (currently in graduate
> school
> > for my masters), and have quite an interest in database design, so after
> > looking into cockroachdb, I'm pretty interested. If possible, I'd like to
> > help you guys out. Any pointers on where I can look to get started?
> >
> > -Mitchell Tracy
> >
> > On Wed, Feb 3, 2016 at 9:51 AM, Cuong Do <notifications@github.com>
> wrote:
> >
> > > Review status: 17 of 19 files reviewed at latest revision, 4 unresolved
> > > discussions, some commit checks failed.
> > > ------------------------------
> > >
> > > *sql/executor.go, line 552 [r4]
> > > <
> >
> https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YWEPeKz_Sngq8NKGx:-K9bXaEf8T2EpwwOtYU_:-499146009
> > >
> > > (raw file
> > > <
> >
> https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/executor.go#L552
> > >):*
> > > Done.
> > > ------------------------------
> > >
> > > *sql/pgwire/server.go, line 228 [r4]
> > > <
> >
> https://reviewable.io:443/reviews/cockroachdb/cockroach/4031#-K9YX-QK5lKXL5239UhI:-K9bbObJffst_ouMHI50:-499146009
> > >
> > > (raw file
> > > <
> >
> https://github.com/cockroachdb/cockroach/blob/a2fa54f16b64f573f2e5a0076109f05eff9adda4/sql/pgwire/server.go#L228
> > >):*
> > > Done.
> > > ------------------------------
> > >
> > > *Comments from the review on Reviewable.io
> > > <https://reviewable.io:443/reviews/cockroachdb/cockroach/4031>*
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179274342>
> > > .
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179313567>
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4031#issuecomment-179333463>
> .
>
